### ISSUE TITLE
Calling kernel with array arguments after having called kernel with texture arguments causes exeption

### ISSUE THREAD

**Issue by nowachri-de** [gpujs/gpu.js#513]
## *What* is wrong?
Calling a kernel with array arguments after having called the same kernel with texture arguments causes exeption. 

If the way I use the kernel is correct, then I think this is a bug report. In case that I am using the
kernel in a wrong way, then I would be happy about a hint on how the propper way of using the kernel is :-)

## *Where* does it happen?
Running gpu.js on nodes version 10.16.3
Platform: | Win32
Chrome/77.0.3865.90 
webgl2
WebGL 2.0 (OpenGL ES 3.0 Chromium)
WebGL GLSL ES 3.00 (OpenGL ES GLSL ES 3.0 Chromium)
WebKit
WebKit WebGL
ANGLE (Intel(R) HD Graphics 520 Direct3D11 vs_5_0 ps_5_0)

## *How* do we replicate the issue?
Execute the code attached below

## *How* important is this (1-5)?
For me it is quit important. I have investigated current and old issues. So I give it a 4
## Expected behavior (i.e. solution)
It should be possible to pass a valid argument (array or texture) to a kernel. If the same kernel is called another time and the argument type has changed, let´s say from texture to array, the kernel should be able to handle the changed argument type in case kernel.dynamicArguments(true) is set.

## Other Comments
The problem occurs at the end of the code, please see comments in code
```
"use strict";
const { GPU } = require('gpu.js');
const gpu = new GPU();

//Dimensions of matrices
let D = 2;

//Standard matrix multiplication
const multiplyMatrix = gpu.createKernel(function (a, b) {
    let sum = 0;
    for (let i = 0; i < this.constants.size; i++) {
        sum += a[this.thread.y][i] * b[i][this.thread.x];
    }
    return sum;
}, {
    constants: { size: D },
    loopMaxIterations: D,
});
multiplyMatrix.setDynamicArguments(true);
multiplyMatrix.setOutput([D, D]);
multiplyMatrix.setPipeline(true);
multiplyMatrix.setImmutable(true);

//Generate Random Numbers
function randomNumbersAtScale2D(x, y, divisor) {
    var matrix = []; // Initialize array
    var i;
    for (i = 0; i < y; i++) {
        matrix[i] = []; // Initialize inner array
        for (var j = 0; j < x; j++) { // i++ needs to be j++
            matrix[i][j] = (Math.random() / divisor);
        }
    }
    return matrix;
}

//convert 2D array to texture
const matrix2Texture = gpu.createKernel(function (dataIn) {
    return dataIn[this.thread.y][this.thread.x];
});

matrix2Texture.setDynamicArguments(true);
matrix2Texture.setDynamicOutput(true);
matrix2Texture.setPipeline(true);
matrix2Texture.setOutput([D, D]);

//create 2D matrix with D dimensions
let a = randomNumbersAtScale2D(D, D, 1);

//create 2D matrix with D dimensions
let b = randomNumbersAtScale2D(D, D, 1);

//test matrix2Texture --> values of a must be the same as values of matrix2Texture(a).toArray()
//rounding errors are O.K
console.log(a);
console.log(matrix2Texture(a).toArray());

//!!!!!!!!!!!! THIS gives a wrong result !!!!!!!!!!!!!!!!!!!!
console.log(multiplyMatrix(matrix2Texture(a), matrix2Texture(b)).toArray());

//!!!!!!!!!!! THIS throws an exception
console.log(multiplyMatrix(a,b).toArray());
```

**Comment by robertleeplummerjr** [comment]
Debugging this now, it seems that the textures for arguments in the special case of them being textures (and with some other qualification that I've not yet found) they are reversed...  Weird.  But still digging.

**Comment by robertleeplummerjr** [comment]
Ah, I see now the issue.  The problem is that the output texture from the function you are calling is being overwritten by the last call to it.  This is fixed by doing the following:

```js
matrix2Texture.setImmutable(true);
```

This will, however, cause a bit higher memory usage, fyi.  What we really need (**NOTE: this is hypothetical**) is a means of copying the texture just in time, like this:

```js
console.log(multiplyMatrix(matrix2Texture(a).copy(), matrix2Texture(b).copy()).toArray());
```

**Comment by nowachri-de** [comment]
Hello Robert,

thanx for your response. I removed the statement matrix2Texture.setImmutable(true) and executed the code again, the result is the same, meaning the last statement still throws an exception.

**Comment by robertleeplummerjr** [comment]
I got you covered, I've added (though a tad changed from above) the new fixes and changes to the API, and will have it pushed up in a few minutes.

I've managed to, after thinking about what we're actually doing with the texture, arrive at `texture.clone()`, over `texture.copy()`.  Yes, `copy` is shorter, but I believe `clone` more covers what we're doing here. Also I found a bug that causes the throw and will have that fixed too.
